primitives/src/transaction.rs

```diff
@@ -89,3 +93,133 @@ impl Version {
+    pub txid: Txid,
+    /// The index of the referenced output in its transaction's vout.
+    pub vout: u32,
+}
+#[cfg(feature = "serde")]
+internals::serde_struct_human_string_impl!(OutPoint, "an OutPoint", txid, vout);
+
+impl OutPoint {
+    /// The number of bytes that an outpoint contributes to the size of a transaction.
+    pub const SIZE: usize = 32 + 4; // The serialized lengths of txid and vout.
+
+    /// Creates a new [`OutPoint`].
+    #[inline]
+    pub const fn new(txid: Txid, vout: u32) -> OutPoint { OutPoint { txid, vout } }
+
+    /// Creates a "null" `OutPoint`.
+    ///
+    /// This value is used for coinbase transactions because they don't have any previous outputs.
+    #[inline]
+    pub fn null() -> OutPoint { OutPoint { txid: Txid::all_zeros(), vout: u32::MAX } }
+
+    /// Checks if an `OutPoint` is "null".
+    #[inline]
+    pub fn is_null(&self) -> bool { *self == OutPoint::null() }
+}
+
+impl Default for OutPoint {
+    fn default() -> Self { OutPoint::null() }
+}
+
+impl fmt::Display for OutPoint {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}:{}", self.txid, self.vout)
+    }
+}
+
+#[cfg(feature = "alloc")]       // Because of `parse::ParseIntError`.
+impl core::str::FromStr for OutPoint {
+    type Err = ParseOutPointError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.len() > 75 {
+            // 64 + 1 + 10
+            return Err(ParseOutPointError::TooLong);
+        }
+        let find = s.find(':');
+        if find.is_none() || find != s.rfind(':') {
+            return Err(ParseOutPointError::Format);
+        }
+        let colon = find.unwrap();
+        if colon == 0 || colon == s.len() - 1 {
+            return Err(ParseOutPointError::Format);
+        }
+        Ok(OutPoint {
+            txid: s[..colon].parse().map_err(ParseOutPointError::Txid)?,
+            vout: parse_vout(&s[colon + 1..])?,
+        })
+    }
+}
+
+/// Parses a string-encoded transaction index (vout).
+///
+/// Does not permit leading zeroes or non-digit characters.
+#[cfg(feature = "alloc")]       // Because of `parse::ParseIntError`.
+fn parse_vout(s: &str) -> Result<u32, ParseOutPointError> {
+    if s.len() > 1 {
+        let first = s.chars().next().unwrap();
+        if first == '0' || first == '+' {
+            return Err(ParseOutPointError::VoutNotCanonical);
+        }
+    }
+    parse::int(s).map_err(ParseOutPointError::Vout)
+}
+
+/// An error in parsing an OutPoint.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+#[cfg(feature = "alloc")]       // Because of `parse::ParseIntError`.
+pub enum ParseOutPointError {
+    /// Error in TXID part.
+    Txid(hex::HexToArrayError),
+    /// Error in vout part.
+    Vout(parse::ParseIntError),
+    /// Error in general format.
+    Format,
+    /// Size exceeds max.
+    TooLong,
+    /// Vout part is not strictly numeric without leading zeroes.
+    VoutNotCanonical,
```

Comment:
Hmm, this suggests that the encoding is actually standardized.

Reply:
It's not (I checked Core and it appears to not only accept leading zeros, but also `+` and `-`s lol). But this "canonical" requirement is there so that we can write fuzztests that check there is a 1-to-1 mapping between outpoints and their string representations.

I'd really rather not drop the canonicity requirement (which you may be the first person ever to even notice it *was* a requirement) because it'll eliminate that category of fuzztests.

Reply:
Can't we just filter out fuzztests differently?

Reply:
Yes, in principle it's always possible to enumerate all forms of non-canonicity and filter for them (or re-canonicalize serializations, or define a weak notion of "equality").

But it's at least an order of magnitude more work than the "build an object, serialize and parse it, assert equality" type fuzztests that we typically have because they're so easy to write.

Reply:
What really bugs me is the inconsistency among implementations. In extreme cases such things can lead to vulnerabilities (though I have hard time imagining how they'd be abused here). OTOH restricting formats is usually a good thing.